### ISSUE TITLE
charhook.dll cannot be overwritten during reinstallation process

### ISSUE THREAD

**Issue by nvdaTrac** [nvaccess/test_nvdaTracImport#202]
**Reported by vtsaran on 24 Jun 2008 17:47 UTC**
During reinstallation of NVDA, i.e. when overwriting existing installation, the installer is unable to update charhook.dll because it is write-protected, thus causing "abort/ignore/retry" error during the installer. Even if NVDA is forcefully shut down, charhook.dll still stays in memory and does not unlink until the next restart of the computer.

**Comment by nvdaTrac** [comment]
**Modified by vtsaran on 24 Jun 2008 17:49 UTC**

**Comment by nvdaTrac** [comment]
**Comment by vtsaran on 24 Jun 2008 17:56 UTC**
Small correction (since there seems to be no way to edit actual description): the dialog produced by the installer when the write-protected file is encountered is "abort/ignore/retry".

**Comment by nvdaTrac** [comment]
**Comment by jteh on 25 Jun 2008 02:39 UTC**
We have investigated this and found the following:
 * This library does not always fail to unload. On my system, it does seem to be unloaded most of the time.
 * We do this using a global hook installed using !SetWindowsHookEx. Using this mechanism, the system loads the library into other processes; we do not explicitly load it. Thus, we can't explicitly unload it. The documentation notes that the library is not explicitly and immediately unloaded, so one assumes the system does it at its leisure.
 * We may be able to get around this by hooking individual threads instead of setting a global hook. In this case, the documentation notes that the system is supposed to explicitly unload the library when the hook is removed.
 * Nevertheless, we're not certain this will fix the problem in all cases.
 * We really should be using the !InstallLib macro in NSIS to install libraries. Among other things, this handles dll versioning, handles reboots if required and allows us to bypass Windows file protection. This should be done for all dlls we install.
  * This will require that these dlls be installed separately from the rest of the package.

**Comment by nvdaTrac** [comment]
**Comment by vtsaran on 25 Jun 2008 21:28 UTC**
Never heard of the !InstallLib macro; will investigate.

**Comment by nvdaTrac** [comment]
**Comment by jteh on 26 Jun 2008 01:57 UTC**
I have made the appropriate changes to use !InstallLib. Shall commit later today.

**Comment by nvdaTrac** [comment]
**Comment by jteh on 26 Jun 2008 03:53 UTC**
We now use !InstallLib and !UninstallLib as of r2161. This means that the installer will now request a reboot if required when handling libraries. Please test and comment.

**Comment by nvdaTrac** [comment]
**Comment by jteh on 11 Aug 2008 22:59 UTC**
No one has reported any issues with this fix, nor any reoccurrence of this problem.
